File: src/jvm/package/src/doc-examples/java/questpdf/docexamples/TableExamples.java
package questpdf.docexamples;

import com.questpdf.fluent.Document;
import com.questpdf.fluent.TableDescriptor;
import com.questpdf.helpers.Colors;
import com.questpdf.helpers.PageSize;
import com.questpdf.helpers.Placeholders;
import com.questpdf.infrastructure.Color;
import com.questpdf.infrastructure.IContainer;
import com.questpdf.infrastructure.ImageCompressionQuality;
import com.questpdf.infrastructure.ImageFormat;
import com.questpdf.infrastructure.ImageGenerationSettings;
import org.junit.jupiter.api.Test;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.ThreadLocalRandom;
import java.util.function.BiFunction;
import java.util.function.Function;

public class TableExamples extends DocExample {

    @Test
    public void basic() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(500f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .table(table -> {
                            table.columnsDefinition(columns -> {
                                columns.constantColumn(50f);
                                columns.relativeColumn();
                                columns.constantColumn(125f);
                            });

                            table.header(header -> {
                                header.cell().borderBottom(2f).padding(8f).text("#");
                                header.cell().borderBottom(2f).padding(8f).text("Product");
                                header.cell().borderBottom(2f).padding(8f).alignRight().text("Price");
                            });

                            for (var i = 0; i < 6; i++) {
                                // The Java counterpart of C# Math.Round(Random.Shared.NextDouble() * 100, 2).
                                var price = Math.round(ThreadLocalRandom.current().nextDouble() * 100 * 100) / 100.0;

                                table.cell().padding(8f).text(String.valueOf(i + 1));
                                table.cell().padding(8f).text(Placeholders.label());
                                table.cell().padding(8f).alignRight().text("$" + price);
                            }
                        });
                });
            })
            .generateImages(index -> output("table-simple.webp"), settings);
    }

    @Test
    public void cellStyleExample() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(500f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    var weatherIcons = new String[] { "cloudy.svg", "lightning.svg", "pouring.svg", "rainy.svg", "snowy.svg", "windy.svg" };

                    page.content()
                        .table(table -> {
                            table.columnsDefinition(columns -> {
                                columns.relativeColumn();
                                columns.constantColumn(125f);
                                columns.constantColumn(125f);
                            });

                            table.header(header -> {
                                // The C# original routes CellStyle through the Element(Func<IContainer, IContainer>)
                                // overload, which is not bridged; a local Function is the direct equivalent.
                                Function<IContainer, IContainer> cellStyle = container ->
                                    container.background(Colors.Blue.getDarken2())
                                        .defaultTextStyle(style -> style.fontColor(Colors.getWhite()).bold())
                                        .paddingVertical(8f)
                                        .paddingHorizontal(16f);

                                cellStyle.apply(header.cell()).text("Day");
                                cellStyle.apply(header.cell()).alignCenter().text("Weather");
                                cellStyle.apply(header.cell()).alignRight().text("Temp");
                            });

                            for (var i = 0; i < 7; i++) {
                                var weatherIndex = ThreadLocalRandom.current().nextInt(0, weatherIcons.length);

                                var backgroundColor = i % 2 == 0
                                    ? Colors.Blue.getLighten5()
                                    : Colors.Blue.getLighten4();

                                Function<IContainer, IContainer> cellStyle = container ->
                                    container.background(backgroundColor)
                                        .paddingVertical(8f)
                                        .paddingHorizontal(16f);

                                cellStyle.apply(table.cell())
                                    .text(LocalDate.of(2025, 2, 26).plusDays(i).format(DateTimeFormatter.ofPattern("dd MMMM", Locale.ENGLISH)));

                                cellStyle.apply(table.cell()).alignCenter().height(24f)
                                    .svg(resource("WeatherIcons/" + weatherIcons[weatherIndex]));

                                cellStyle.apply(table.cell()).alignRight()
                                    .text(ThreadLocalRandom.current().nextInt(-10, 35) + "°");
                            }
                        });
                });
            })
            .generateImages(index -> output("table-cell-style.webp"), settings);
    }

    @Test
    public void overlappingCells() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(700f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(16f));
                    page.margin(25f);

                    var dayNames = new String[] { "Monday", "Tuesday", "Wednesday", "Thursday", "Friday" };

                    page.content()
                        .border(1f)
                        .borderColor(Colors.Grey.getLighten1())
                        .table(table -> {
                            table.columnsDefinition(columns -> {
                                // hour column
                                columns.constantColumn(60f);

                                // day columns
                                for (var i = 0; i < 5; i++)
                                    columns.relativeColumn();
                            });

                            // even/odd columns background
                            for (var column = 0; column < 7; column++) {
                                var backgroundColor = column % 2 == 0 ? Colors.Grey.getLighten3() : Colors.getWhite();
                                table.cell().column(column).rowSpan(24).background(backgroundColor);
                            }

                            // hours and hour lines
                            for (var hour = 6; hour < 16; hour++) {
                                table.cell().column(1).row(hour)
                                    .paddingVertical(5f).paddingHorizontal(10f).alignRight()
                                    .text(String.valueOf(hour));

                                table.cell().row(hour).columnSpan(6)
                                    .border(1f).borderColor(Colors.Grey.getLighten1()).height(20f);
                            }

                            // dates and day names
                            for (var i = 0; i < 5; i++) {
                                var index = i;

                                table.cell()
                                    .column(i + 2).row(1).padding(5f)
                                    .column(column -> {
                                        column.item().alignCenter().text(String.valueOf(17 + index)).fontSize(24f).bold();
                                        column.item().alignCenter().text(dayNames[index]).light();
                                    });
                            }

                            // standup events
                            for (var i = 1; i <= 4; i++)
                                addEvent(table, i, 8, 1, "Standup", Colors.Blue.getLighten4(), Colors.Blue.getDarken3());

                            // other events
                            addEvent(table, 2, 11, 2, "Interview", Colors.Red.getLighten4(), Colors.Red.getDarken3());
                            addEvent(table, 3, 12, 3, "Demo", Colors.Red.getLighten4(), Colors.Red.getDarken3());
                            addEvent(table, 5, 5, 17, "PTO", Colors.Green.getLighten4(), Colors.Green.getDarken3());
                        });
                });
            })
            .generateImages(index -> output("table-overlapping-cells.webp"), settings);
    }

    private static void addEvent(TableDescriptor table, int day, int hour, int length, String name, Color backgroundColor, Color textColor) {
        table.cell()
            .column(day + 1).row(hour).rowSpan(length)
            .padding(5f).background(backgroundColor).padding(5f)
            .alignCenter().alignMiddle()
            .text(name).fontColor(textColor);
    }

    @Test
    public void manualCellPlacement() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(700f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(16f));
                    page.margin(25f);

                    page.content()
                        .table(table -> {
                            table.columnsDefinition(columns -> {
                                columns.constantColumn(75f);
                                columns.constantColumn(150f);
                                columns.constantColumn(200f);
                                columns.constantColumn(200f);
                            });

                            // The C# original routes the cell styles through the Element(Func<IContainer, IContainer>)
                            // overload, which is not bridged; local Functions are the direct equivalent.
                            BiFunction<IContainer, Color, IContainer> cellStyle = (container, color) ->
                                container.border(1f).background(color).paddingHorizontal(10f).paddingVertical(15f).alignCenter().alignMiddle();

                            Function<IContainer, IContainer> headerCellStyle = container ->
                                cellStyle.apply(container, Colors.Grey.getLighten4());

                            Function<IContainer, IContainer> goodCellStyle = container ->
                                cellStyle.apply(container, Colors.Green.getLighten4()).defaultTextStyle(style -> style.fontColor(Colors.Green.getDarken2()));

                            Function<IContainer, IContainer> badCellStyle = container ->
                                cellStyle.apply(container, Colors.Red.getLighten4()).defaultTextStyle(style -> style.fontColor(Colors.Red.getDarken2()));

                            headerCellStyle.apply(table.cell().row(1).column(3).columnSpan(2))
                                .text("Predicted condition").bold();

                            headerCellStyle.apply(table.cell().row(3).column(1).rowSpan(2)).rotateLeft()
                                .text("Actual\ncondition").bold().alignCenter();

                            headerCellStyle.apply(table.cell().row(2).column(3))
                                .text("Positive (PP)");

                            headerCellStyle.apply(table.cell().row(2).column(4))
                                .text("Negative (PN)");

                            headerCellStyle.apply(table.cell().row(3).column(2)).text("Positive (P)");

                            headerCellStyle.apply(table.cell().row(4).column(2))
                                .text("Negative (N)");

                            goodCellStyle.apply(table.cell().row(3).column(3))
                                .text("True positive (TP)");

                            badCellStyle.apply(table.cell().row(3).column(4))
                                .text("False negative (FN)");

                            badCellStyle.apply(table.cell().row(4).column(3)).text("False positive (FP)");

                            goodCellStyle.apply(table.cell().row(4).column(4)).text("True negative (TN)");
                        });
                });
            })
            .generateImages(index -> output("table-manual-cell-placement.webp"), settings);
    }

    @Test
    public void columnsDefinition() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(700f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(16f));
                    page.margin(25f);

                    page.content()
                        .width(450f)
                        .table(table -> {
                            table.columnsDefinition(columns -> {
                                columns.constantColumn(150f);
                                columns.relativeColumn(2f);
                                columns.relativeColumn(3f);
                            });

                            // The C# original routes CellStyle through the Element(Func<IContainer, IContainer>)
                            // overload, which is not bridged; a local Function is the direct equivalent.
                            Function<IContainer, IContainer> cellStyle = container ->
                                container.border(1f).padding(10f);

                            cellStyle.apply(table.cell().columnSpan(3).background(Colors.Grey.getLighten2()))
                                .text("Total width: 450px");

                            cellStyle.apply(table.cell()).text("Constant: 150px");
                            cellStyle.apply(table.cell()).text("Relative: 2*");
                            cellStyle.apply(table.cell()).text("Relative: 3*");

                            cellStyle.apply(table.cell()).text("150px");
                            cellStyle.apply(table.cell()).text("120px");
                            cellStyle.apply(table.cell()).text("180px");
                        });
                });
            })
            .generateImages(index -> output("table-columns-definition.webp"), settings);
    }

    @Test
    public void headerAndFooter() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(600f, 250f));
                    page.defaultTextStyle(style -> style.fontSize(16f));
                    page.margin(25f);

                    page.content()
                        .border(1f)
                        .borderColor(Colors.Grey.getLighten1())
                        .table(table -> {
                            record NamedPageSize(String name, double width, double height) {}

                            var pageSizes = List.of(
                                new NamedPageSize("Letter (ANSI A)", 8.5, 11.0),
                                new NamedPageSize("Legal", 8.5, 14.0),
                                new NamedPageSize("Ledger (ANSI B)", 11.0, 17.0),
                                new NamedPageSize("Tabloid (ANSI B)", 17.0, 11.0),
                                new NamedPageSize("ANSI C", 22.0, 17.0),
                                new NamedPageSize("ANSI D", 34.0, 22.0),
                                new NamedPageSize("ANSI E", 44.0, 34.0));

                            var inchesToPoints = 72;

                            // The C# original routes the cell styles through the Element(Func<IContainer, IContainer>)
                            // overload, which is not bridged; local Functions are the direct equivalent.
                            // (C# declares backgroundColor as string, relying on implicit Color<->string conversions.)
                            BiFunction<IContainer, Color, IContainer> defaultCellStyle = (container, backgroundColor) ->
                                container
                                    .border(1f)
                                    .borderColor(Colors.Grey.getLighten1())
                                    .background(backgroundColor)
                                    .paddingVertical(5f)
                                    .paddingHorizontal(10f)
                                    .alignCenter()
                                    .alignMiddle();

                            table.columnsDefinition(columns -> {
                                columns.relativeColumn();

                                columns.constantColumn(80f);
                                columns.constantColumn(80f);

                                columns.constantColumn(80f);
                                columns.constantColumn(80f);
                            });

                            table.header(header -> {
                                // please be sure to call the 'header' handler!

                                // you can extend existing styles by creating additional methods
                                Function<IContainer, IContainer> cellStyle = container ->
                                    defaultCellStyle.apply(container, Colors.Grey.getLighten3());

                                cellStyle.apply(header.cell().rowSpan(2)).extendHorizontal().alignLeft()
                                    .text("Document type").bold();

                                cellStyle.apply(header.cell().columnSpan(2)).text("Inches").bold();
                                cellStyle.apply(header.cell().columnSpan(2)).text("Points").bold();

                                cellStyle.apply(header.cell()).text("Width");
                                cellStyle.apply(header.cell()).text("Height");

                                cellStyle.apply(header.cell()).text("Width");
                                cellStyle.apply(header.cell()).text("Height");
                            });

                            for (var pageSize : pageSizes) {
                                Function<IContainer, IContainer> cellStyle = container ->
                                    defaultCellStyle.apply(container, Colors.getWhite()).showOnce();

                                cellStyle.apply(table.cell()).extendHorizontal().alignLeft().text(pageSize.name());

                                // inches
                                cellStyle.apply(table.cell()).text(formatNumber(pageSize.width()));
                                cellStyle.apply(table.cell()).text(formatNumber(pageSize.height()));

                                // points
                                cellStyle.apply(table.cell()).text(formatNumber(pageSize.width() * inchesToPoints));
                                cellStyle.apply(table.cell()).text(formatNumber(pageSize.height() * inchesToPoints));
                            }
                        });
                });
            })
            .generateImages(index -> output("table-header-and-footer-" + index + ".webp"), settings);
    }

    /**
     * The string .NET produces for a double passed to Text(object): integral values
     * print without a fractional part (e.g. 612), non-integral ones with it (e.g. 8.5).
     */
    private static String formatNumber(double value) {
        return value == Math.floor(value) ? String.valueOf((long) value) : String.valueOf(value);
    }
}

File: src/jvm/package/src/doc-examples/kotlin/questpdf/docexamples/TableExamples.kt
package questpdf.docexamples

import com.questpdf.fluent.Document
import com.questpdf.helpers.Colors
import com.questpdf.helpers.PageSize
import com.questpdf.helpers.Placeholders
import com.questpdf.infrastructure.Color
import com.questpdf.infrastructure.IContainer
import com.questpdf.infrastructure.ImageCompressionQuality
import com.questpdf.infrastructure.ImageFormat
import com.questpdf.infrastructure.ImageGenerationSettings
import org.junit.jupiter.api.Test
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.Locale
import kotlin.math.floor
import kotlin.math.round
import kotlin.random.Random

class TableExamples : DocExample() {

    @Test
    fun basic() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(500f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .table {
                            columnsDefinition {
                                constantColumn(50f)
                                relativeColumn()
                                constantColumn(125f)
                            }

                            header {
                                cell().borderBottom(2f).padding(8f).text("#")
                                cell().borderBottom(2f).padding(8f).text("Product")
                                cell().borderBottom(2f).padding(8f).alignRight().text("Price")
                            }

                            for (i in 0 until 6) {
                                // The Kotlin counterpart of C# Math.Round(Random.Shared.NextDouble() * 100, 2).
                                val price = round(Random.nextDouble() * 100 * 100) / 100

                                cell().padding(8f).text("${i + 1}")
                                cell().padding(8f).text(Placeholders.label())
                                cell().padding(8f).alignRight().text("\$$price")
                            }
                        }
                }
            }
            .generateImages({ output("table-simple.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun cellStyleExample() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(500f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    val weatherIcons = arrayOf("cloudy.svg", "lightning.svg", "pouring.svg", "rainy.svg", "snowy.svg", "windy.svg")

                    content()
                        .table {
                            columnsDefinition {
                                relativeColumn()
                                constantColumn(125f)
                                constantColumn(125f)
                            }

                            header {
                                // The C# original routes CellStyle through the Element(Func<IContainer, IContainer>)
                                // overload, which is not bridged; an extension function is the direct equivalent.
                                fun IContainer.cellStyle(): IContainer {
                                    return background(Colors.Blue.Darken2)
                                        .defaultTextStyle { fontColor(Colors.White).bold() }
                                        .paddingVertical(8f)
                                        .paddingHorizontal(16f)
                                }

                                cell().cellStyle().text("Day")
                                cell().cellStyle().alignCenter().text("Weather")
                                cell().cellStyle().alignRight().text("Temp")
                            }

                            for (i in 0 until 7) {
                                val weatherIndex = Random.nextInt(0, weatherIcons.size)

                                fun IContainer.cellStyle(): IContainer {
                                    val backgroundColor = if (i % 2 == 0)
                                        Colors.Blue.Lighten5
                                    else
                                        Colors.Blue.Lighten4

                                    return background(backgroundColor)
                                        .paddingVertical(8f)
                                        .paddingHorizontal(16f)
                                }

                                cell().cellStyle()
                                    .text(LocalDate.of(2025, 2, 26).plusDays(i.toLong()).format(DateTimeFormatter.ofPattern("dd MMMM", Locale.ENGLISH)))

                                cell().cellStyle().alignCenter().height(24f)
                                    .svg(resource("WeatherIcons/${weatherIcons[weatherIndex]}"))

                                cell().cellStyle().alignRight()
                                    .text("${Random.nextInt(-10, 35)}°")
                            }
                        }
                }
            }
            .generateImages({ output("table-cell-style.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun overlappingCells() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(700f, 1000f))
                    defaultTextStyle { fontSize(16f) }
                    margin(25f)

                    val dayNames = arrayOf("Monday", "Tuesday", "Wednesday", "Thursday", "Friday")

                    content()
                        .border(1f)
                        .borderColor(Colors.Grey.Lighten1)
                        .table {
                            columnsDefinition {
                                // hour column
                                constantColumn(60f)

                                // day columns
                                repeat(5) {
                                    relativeColumn()
                                }
                            }

                            // even/odd columns background
                            for (column in 0 until 7) {
                                val backgroundColor = if (column % 2 == 0) Colors.Grey.Lighten3 else Colors.White
                                cell().column(column.toUInt()).rowSpan(24u).background(backgroundColor)
                            }

                            // hours and hour lines
                            for (hour in 6 until 16) {
                                cell().column(1u).row(hour.toUInt())
                                    .paddingVertical(5f).paddingHorizontal(10f).alignRight()
                                    .text("$hour")

                                cell().row(hour.toUInt()).columnSpan(6u)
                                    .border(1f).borderColor(Colors.Grey.Lighten1).height(20f)
                            }

                            // dates and day names
                            for (i in 0 until 5) {
                                cell()
                                    .column(i.toUInt() + 2u).row(1u).padding(5f)
                                    .column {
                                        item().alignCenter().text("${17 + i}").fontSize(24f).bold()
                                        item().alignCenter().text(dayNames[i]).light()
                                    }
                            }

                            fun addEvent(day: UInt, hour: UInt, length: UInt, name: String, backgroundColor: Color, textColor: Color) {
                                cell()
                                    .column(day + 1u).row(hour).rowSpan(length)
                                    .padding(5f).background(backgroundColor).padding(5f)
                                    .alignCenter().alignMiddle()
                                    .text(name).fontColor(textColor)
                            }

                            // standup events
                            for (i in 1..4)
                                addEvent(i.toUInt(), 8u, 1u, "Standup", Colors.Blue.Lighten4, Colors.Blue.Darken3)

                            // other events
                            addEvent(2u, 11u, 2u, "Interview", Colors.Red.Lighten4, Colors.Red.Darken3)
                            addEvent(3u, 12u, 3u, "Demo", Colors.Red.Lighten4, Colors.Red.Darken3)
                            addEvent(5u, 5u, 17u, "PTO", Colors.Green.Lighten4, Colors.Green.Darken3)
                        }
                }
            }
            .generateImages({ output("table-overlapping-cells.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun manualCellPlacement() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(700f, 1000f))
                    defaultTextStyle { fontSize(16f) }
                    margin(25f)

                    content()
                        .table {
                            columnsDefinition {
                                constantColumn(75f)
                                constantColumn(150f)
                                constantColumn(200f)
                                constantColumn(200f)
                            }

                            // The C# original routes the cell styles through the Element(Func<IContainer, IContainer>)
                            // overload, which is not bridged; extension functions are the direct equivalent.
                            fun IContainer.cellStyle(color: Color): IContainer =
                                border(1f).background(color).paddingHorizontal(10f).paddingVertical(15f).alignCenter().alignMiddle()

                            fun IContainer.headerCellStyle(): IContainer =
                                cellStyle(Colors.Grey.Lighten4)

                            fun IContainer.goodCellStyle(): IContainer =
                                cellStyle(Colors.Green.Lighten4).defaultTextStyle { fontColor(Colors.Green.Darken2) }

                            fun IContainer.badCellStyle(): IContainer =
                                cellStyle(Colors.Red.Lighten4).defaultTextStyle { fontColor(Colors.Red.Darken2) }

                            cell().row(1u).column(3u).columnSpan(2u)
                                .headerCellStyle()
                                .text("Predicted condition").bold()

                            cell().row(3u).column(1u).rowSpan(2u)
                                .headerCellStyle().rotateLeft()
                                .text("Actual\ncondition").bold().alignCenter()

                            cell().row(2u).column(3u)
                                .headerCellStyle()
                                .text("Positive (PP)")

                            cell().row(2u).column(4u)
                                .headerCellStyle()
                                .text("Negative (PN)")

                            cell().row(3u).column(2u)
                                .headerCellStyle().text("Positive (P)")

                            cell().row(4u).column(2u)
                                .headerCellStyle()
                                .text("Negative (N)")

                            cell()
                                .row(3u).column(3u).goodCellStyle()
                                .text("True positive (TP)")

                            cell()
                                .row(3u).column(4u).badCellStyle()
                                .text("False negative (FN)")

                            cell().row(4u).column(3u)
                                .badCellStyle().text("False positive (FP)")

                            cell().row(4u).column(4u)
                                .goodCellStyle().text("True negative (TN)")
                        }
                }
            }
            .generateImages({ output("table-manual-cell-placement.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun columnsDefinition() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(700f, 1000f))
                    defaultTextStyle { fontSize(16f) }
                    margin(25f)

                    content()
                        .width(450f)
                        .table {
                            columnsDefinition {
                                constantColumn(150f)
                                relativeColumn(2f)
                                relativeColumn(3f)
                            }

                            // The C# original routes CellStyle through the Element(Func<IContainer, IContainer>)
                            // overload, which is not bridged; an extension function is the direct equivalent.
                            fun IContainer.cellStyle(): IContainer =
                                border(1f).padding(10f)

                            cell().columnSpan(3u)
                                .background(Colors.Grey.Lighten2).cellStyle()
                                .text("Total width: 450px")

                            cell().cellStyle().text("Constant: 150px")
                            cell().cellStyle().text("Relative: 2*")
                            cell().cellStyle().text("Relative: 3*")

                            cell().cellStyle().text("150px")
                            cell().cellStyle().text("120px")
                            cell().cellStyle().text("180px")
                        }
                }
            }
            .generateImages({ output("table-columns-definition.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun headerAndFooter() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(600f, 250f))
                    defaultTextStyle { fontSize(16f) }
                    margin(25f)

                    content()
                        .border(1f)
                        .borderColor(Colors.Grey.Lighten1)
                        .table {
                            val pageSizes = listOf(
                                Triple("Letter (ANSI A)", 8.5, 11.0),
                                Triple("Legal", 8.5, 14.0),
                                Triple("Ledger (ANSI B)", 11.0, 17.0),
                                Triple("Tabloid (ANSI B)", 17.0, 11.0),
                                Triple("ANSI C", 22.0, 17.0),
                                Triple("ANSI D", 34.0, 22.0),
                                Triple("ANSI E", 44.0, 34.0),
                            )

                            val inchesToPoints = 72

                            // The C# original routes the cell styles through the Element(Func<IContainer, IContainer>)
                            // overload, which is not bridged; extension functions are the direct equivalent.
                            // (C# declares backgroundColor as string, relying on implicit Color<->string conversions.)
                            fun IContainer.defaultCellStyle(backgroundColor: Color): IContainer {
                                return border(1f)
                                    .borderColor(Colors.Grey.Lighten1)
                                    .background(backgroundColor)
                                    .paddingVertical(5f)
                                    .paddingHorizontal(10f)
                                    .alignCenter()
                                    .alignMiddle()
                            }

                            columnsDefinition {
                                relativeColumn()

                                constantColumn(80f)
                                constantColumn(80f)

                                constantColumn(80f)
                                constantColumn(80f)
                            }

                            header {
                                // please be sure to call the 'header' handler!

                                // you can extend existing styles by creating additional methods
                                fun IContainer.cellStyle(): IContainer =
                                    defaultCellStyle(Colors.Grey.Lighten3)

                                cell().rowSpan(2u).cellStyle().extendHorizontal().alignLeft()
                                    .text("Document type").bold()

                                cell().columnSpan(2u).cellStyle().text("Inches").bold()
                                cell().columnSpan(2u).cellStyle().text("Points").bold()

                                cell().cellStyle().text("Width")
                                cell().cellStyle().text("Height")

                                cell().cellStyle().text("Width")
                                cell().cellStyle().text("Height")
                            }

                            for ((name, width, height) in pageSizes) {
                                fun IContainer.cellStyle(): IContainer =
                                    defaultCellStyle(Colors.White).showOnce()

                                cell().cellStyle().extendHorizontal().alignLeft().text(name)

                                // inches
                                cell().cellStyle().text(formatNumber(width))
                                cell().cellStyle().text(formatNumber(height))

                                // points
                                cell().cellStyle().text(formatNumber(width * inchesToPoints))
                                cell().cellStyle().text(formatNumber(height * inchesToPoints))
                            }
                        }
                }
            }
            .generateImages({ index -> output("table-header-and-footer-$index.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    /**
     * The string .NET produces for a double passed to Text(object): integral values
     * print without a fractional part (e.g. 612), non-integral ones with it (e.g. 8.5).
     */
    private fun formatNumber(value: Double): String =
        if (value == floor(value)) value.toLong().toString() else value.toString()
}
